packages/aws-client/src/lib/S3MultipartUploads.ts

```diff
@@ -0,0 +1,150 @@
+// Utility functions to help with S3 multi-part uploads
+
+import range from 'lodash/range';
+import { s3 } from '../services';
+
+type Chunk = {
+  start: number,
+  end: number
+};
+
+const MB = 1024 * 1024;
+const GB = 1024 * MB;
+
+// Each part of a multi-part copy needs to specify a byte range to be copied.
+// This byte range has a starting byte and an ending byte (inclusive) that makes
+// up the part. The maximum allowed chunk size is 5368709120 bytes.
+//
+// This function takes a file size and an optional maxSize. It returns an array
+// of objects, each containing a `start` and an `end` value. These will make up
+// the ranges of the multi-part copy.
+//
+// From anecdotal testing, a chunk size of 250 MB seems to perform fairly well.
+//
+// https://docs.aws.amazon.com/AmazonS3/latest/API/API_UploadPartCopy.html
+export const createMultipartChunks = (
+  objectSize: number,
+  maxChunkSize = 250 * MB
+): Chunk[] => {
+  if (maxChunkSize > (5 * GB)) {
+    throw new Error('maxChunkSize can be no greater than 5368709120');
+  }
+
+  const lastChunkSize = objectSize % maxChunkSize;
+
+  // Build the list of full-size chunks
+  const chunks = range(0, objectSize - lastChunkSize, maxChunkSize)
+    .map((start) => {
+      const end = start + maxChunkSize - 1;
+      return { start, end };
+    });
+
+  // If necessary build the last, not-full-size chunk
+  if (lastChunkSize !== 0) {
+    const start = objectSize - lastChunkSize;
+    const end = objectSize - 1;
+    chunks.push({ start, end });
+  }
+
+  return chunks;
+};
```

Comment:
This is a small amount of code, so I'm perfectly fine with the additional verbosity, so the rest of this comment is for optional perusal.

BTW, glad to see you converted this file to TypeScript, and also that you found a good default chunk size for good performance. Looks like AWS nicely parallelizes the part uploads.

I agree that a bit of verbosity can indeed be preferable when it greatly clarifies things.

I think it's debatable as to whether that's the case here. I arguably (arguing with myself) find the code below to be more clear, but that's perhaps because I wrote it. Notice that it also avoids checking `maxChunkSize`, leaving `uploadPartCopy` to do the complaining, if necessary.

```javascript
export const createMultipartChunks = (
  objectSize: number,
  maxChunkSize = 250 * MB
): Chunk[] =>
  range(0, objectSize, maxChunkSize).map((start) => ({
    start,
    end: Math.min(start + maxChunkSize, objectSize) - 1
  }));
```

Reply:
Assuming the tests pass, I'll to it this way.